### ISSUE TITLE
general strategies for checking the validity of entries (Value-field)

### ISSUE THREAD

**Issue by LinguList** [glottobank/lexibank-data#17]
I just realized that this issue may occur across more datasets:

* missing entries are often marked as "?", but sometimes just left out "", yet most checks only assume that it is "?"

As a general idea, to make sure that a Value is valid, we might think of a function to be applied to all data, making sure that the value:

* is a valid string of length 1, 
* is not multiple strings in one field (various separators are there, etc.)

and potentially more things.

The newly added function (from lingpy, recent version on github) "clean_sequence", which can be imported from pylexibank.lingpy_util has provided quite useful checks so far, as it strikes if entries are empty, but also if an entry is only a bracket, such as, e.g., (???). Having a general strategy would surely help us, but maybe we'll need to derive it slowly from getting more and more experience iwth the data. 

Examples are in the modified init-files for wwua and also gbd.

**Comment by xrotwang** [comment]
Making sure it's a string of length > 0 by default is ok with me. More elaborate checks should be implemented using a validator function. Maybe there should be support for registration of such functions either in `pylexibank.dataset.CldfDataset` or in `pycldf.dataset.Dataset`. I'm leaning towards the former, since the latter should rather add support for enforcing [datatype specifications](https://www.w3.org/TR/2015/REC-tabular-metadata-20151217/#datatypes) as described in w3's model for tabular metadata.

**Comment by SimonGreenhill** [comment]
Good idea, but strings of length == 1 should also be checked for common 'null' or empty values e.g. 
"-", "?", "ø", or "Ø". 

Re: validators -- yes, but I suspect we'll also need to be able to have dataset specific validators too. So it'd be good to have a subclassable object that has a specific pattern for validation. Dataset specific validators could subclass `pylexibank.dataset.CldfDataset.Validator` and extend it in a validators.py file in each dataset's directory? 

This also brings up a question about what to do with 'invalid' data? do we *not* keep it? or do we flag it as problematic for reason 'x'?

**Comment by LinguList** [comment]
in my bad-ass cleaning procedures, which I usually don't share, I remove whitespace, the characters mentioned by @SimonGreenhill (but beware of the ø, which might be a vowel), but there are many more problematic instances, like "(xxx)", or numbers, etc.

So having some generical, individually expandable validation function would be good. I'm thinking of a workflow like this:

* use the default validator
* check for the clean_string to get a segmented form
* compare problematic entries (those, were segmentation fails, or similar)
* adjust the validator (like setting -666 as pointing to missing data, as in STARLING, etc.)

I would have a couple of ideas for checking, but if one of the two of you could go ahead to make the initial base-validator, it would be easier for me to bring them up, as I don't know how to optimally integrate this in either of the classes.

**Comment by xrotwang** [comment]
regarding validators: My current idea is as follows: `CldfDataset` gets a new keyword argument `validators`, which is expected to be a dictionary of callables. Default validators can then be added via
```python
validators.setdefault(...)
```
and custom validators can include default functionality by simply calling the default functions in their own code.

These validators would be called in `CldfDataset.add_row`. 

Validator functions must have the signature
```python
def validator(row):
    pass
```
thus a validator has access to all fields of a row - so one could simply provide one validator, which checks all fields ...

Regarding semantics, I think we'd need three states:
- validator returns `True`: add the row
- validator return `False`: the row is not added
- validator raises exception: something unexpected happened, which has to be dealt with before calling `add_row`.

OK?

**Comment by LinguList** [comment]
yep, seems like a good idea to me.

Another question, but ignore it if it leads too far, but I have been regularly encountering it, is the problem of excessive synonymy in the rows of a given dataset. I can handle this with python workarounds and in lingpy, but might it be an idea to think of a procedure that takes only the first of three synonymous examples and stores the rest in a field, naming it as "entry-in-source" or the like? 

Note that we have a similar situation already, as the current clean_sequence function splits a multi-entry cell (say, with "/" as separator), but we usually only take the first of the entries that were split for validation, ignoring the rest.

But if we try to do this more regularly, it might be a problem for the validator logic with true/false (unless we manage to handle it in the exception).

**Comment by xrotwang** [comment]
Hm. The only useful cut-off I can think of is 1, right? Why would one want to allow three synonyms, rather than four or whatever? So I think the best we can do is 
- preserve order in the rows, 
- hope that data creators typically list most important/most frequent/most anything first (and that "anything" works for our analysis),
- filter out any more synonyms before analysis.

**Comment by LinguList** [comment]
yep, I'd do the same, and if we make sure to have the alternative entries preserved (say a field with "variants"), scholars who work in concrete on the data could just pick up.

At least that's what I just tested when working a bit on the Dogon data, where excessive synonymy is a result of different fieldworkers whose data has been combined, I suspect. Although I know that classical linguists really love to preserve all this variance and seem to even like it to have so many possibilities, I know it is really problematic for computer approaches, but also for inspecting data. 

We might discuss whether we suppress synonymy pragmatically already from the validators, or whether we instead leave it to be handled on a per-dataset-decision.

In any way we will have some problems if excessive synonymy is also annotated for cognacy. Then we have words in the same language being cognate with themselves, partially, we have partial cognates, etc., and it get's quickly really messy (some ABVD languages had five or six synonyms, when we last checked the app, or it was due to merger of different sources).
